File: 155_MinStack.cpp
//53.28%
class MinStack {
public:
    /** initialize your data structure here. */
    MinStack() {
        
    }
    
    void push(int x) {
        objs.push_back(x);
		if(minObjs.empty() || minObjs[minObjs.size() - 1] >= x)
			minObjs.push_back(x);
    }
    
    void pop() {
        if(objs.empty())
			return;
		if(objs[objs.size() - 1] == minObjs[minObjs.size() - 1])
			minObjs.pop_back();
		objs.pop_back();
    }
    
    int top() {
        if(objs.empty())
			exit(-1);
		return objs[objs.size() - 1];
    }
    
    int getMin() {
        if(objs.empty())
			exit(-1);
		return minObjs[minObjs.size() - 1];
    }
private:
	vector<int> objs;
	vector<int> minObjs;
};

//A Better solution from LeetCode which use two Stacks
class MinStack {
private:
    stack<int> s1;
    stack<int> s2;
public:
    void push(int x) {
	    s1.push(x);
	    if (s2.empty() || x <= getMin())  s2.push(x);	    
    }
    void pop() {
	    if (s1.top() == getMin())  s2.pop();
	    s1.pop();
    }
    int top() {
	    return s1.top();
    }
    int getMin() {
	    return s2.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 155_MinStack.py
# 94.64%
class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.minStack = []

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        self.stack.append(x)
        if not self.minStack or self.minStack[-1] >= x:
            self.minStack.append(x)

    def pop(self):
        """
        :rtype: void
        """
        if not self.stack:
            return
        last = self.stack.pop()
        if self.minStack[-1] == last:
            self.minStack.pop()
        

    def top(self):
        """
        :rtype: int
        """
        if not self.stack:
            return
        return self.stack[-1]

    def getMin(self):
        """
        :rtype: int
        """
        if not self.minStack:
            return
        return self.minStack[-1]

# A better solution than mine from LeetCode which use only one stack list
def __init__(self):
    """
    initialize your data structure here.
    """
    self.stack= []

def push(self, x):
    """
    :type x: int
    :rtype: nothing
    """
    if not self.stack:self.stack.append((x,x)) 
    else:
       self.stack.append((x,min(x,self.stack[-1][1])))

def pop(self):
    """
    :rtype: nothing
    """
    if self.stack: self.stack.pop()

def top(self):
    """
    :rtype: int
    """
    if self.stack: return self.stack[-1][0]
    else: return None

def getMin(self):
    """
    :rtype: int
    """
    if self.stack: return self.stack[-1][1]
    else: return None
# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()